onnxruntime/test/framework/ort_model_only_test.cc

```diff
@@ -0,0 +1,223 @@
+  }
+}
+
+#if !defined(ORT_MINIMAL_BUILD)
+TEST(OrtModelOnlyTests, SerializeToOrtFormat) {
+  const auto output_file = ORT_TSTR("ort_github_issue_4031.onnx.ort");
+  SessionOptions so;
+  so.session_logid = "SerializeToOrtFormat";
+  so.optimized_model_filepath = output_file;
+  // not strictly necessary - type should be inferred from the filename
+  so.AddConfigEntry(ORT_SESSION_OPTIONS_CONFIG_SAVE_MODEL_FORMAT, "ORT");
+
+  InferenceSessionGetGraphWrapper session_object{so, GetEnvironment()};
+
+  // create .ort file during Initialize due to values in SessionOptions
+  ASSERT_STATUS_OK(session_object.Load(ORT_TSTR("testdata/ort_github_issue_4031.onnx")));
+  ASSERT_STATUS_OK(session_object.Initialize());
+
+  // create inputs
+  OrtValue ml_value;
+  CreateMLValue<float>(TestCPUExecutionProvider()->GetAllocator(0, OrtMemTypeDefault), {1}, {123.f},
+                       &ml_value);
+  NameMLValMap feeds;
+  feeds.insert(std::make_pair("state_var_in", ml_value));
+
+  // prepare outputs
+  std::vector<std::string> output_names{"state_var_out"};
+  std::vector<OrtValue> fetches;
+
+  ASSERT_STATUS_OK(session_object.Run(feeds, output_names, &fetches));
+
+  SessionOptions so2;
+  so.session_logid = "LoadOrtFormat";
+  // not strictly necessary - type should be inferred from the filename, but to be sure we're testing what we
+  // think we're testing set it.
+  so.AddConfigEntry(ORT_SESSION_OPTIONS_CONFIG_LOAD_MODEL_FORMAT, "ORT");
+
+  // load serialized version
+  InferenceSessionGetGraphWrapper session_object2{so2, GetEnvironment()};
+  ASSERT_STATUS_OK(session_object2.Load(output_file));
+  ASSERT_STATUS_OK(session_object2.Initialize());
+
+  // compare contents on Graph instances
+  const auto& graph = session_object.GetGraph();
+  const auto& graph2 = session_object2.GetGraph();
+
+  const auto& session_state = session_object.GetSessionState();
+  const auto& session_state2 = session_object2.GetSessionState();
+
+  const auto& i1 = session_state.GetInitializedTensors();
+  const auto& i2 = session_state2.GetInitializedTensors();
+  ASSERT_EQ(i1.size(), i2.size());
+
+  for (const auto& pair : i1) {
+    auto iter = i2.find(pair.first);
+    ASSERT_NE(iter, i2.cend());
+
+    const OrtValue& left = pair.second;
+    const OrtValue& right = iter->second;
+    CompareTensors(left, right);
+  }
+
+  // check all node args are fine
+  for (const auto& input : graph.GetInputsIncludingInitializers()) {
+    const auto& left = *graph.GetNodeArg(input->Name());
+    const auto* right = graph2.GetNodeArg(input->Name());
+    ASSERT_TRUE(right != nullptr);
+
+    const auto& left_proto = left.ToProto();
+    const auto& right_proto = right->ToProto();
+    CompareValueInfos(left_proto, right_proto);
+  }
+
+  for (const auto& left : graph.Nodes()) {
+    const auto* right = graph2.GetNode(left.Index());
+    ASSERT_TRUE(right != nullptr);
+    const auto& left_outputs = left.OutputDefs();
+    const auto& right_outputs = right->OutputDefs();
```

Comment:
Verify outputs only?

Reply:
Verifies graph inputs and all node outputs. That should cover all values that involved type/shape inferencing. Could also do the NodeArg instances for initializers although we did already check all the initializer content.